SEP-0012.md

```diff
@@ -0,0 +1,343 @@
+This will hopefully allow future iteration on the implementation without breaking the functionality of the `NDCube` class.
+
+```python
+
+class ExtraCoordsABC(abc.ABC):
+    """
+    A representation of additional world coordinates associated with pixel axes.
+
+    ExtraCoords can be initialised by either specifying a
+    `~astropy.wcs.wcsapi.LowLevelWCS` object and a ``mapping``, or it can be
+    built up by specifying one or more lookup tables.
+
+    Parameters
+    ----------
+    wcs
+        The WCS specifying the extra coordinates.
+    mapping
+       The mapping between the array dimensions and pixel dimensions in the
+       extra coords object. This is an iterable of ``(array_dimension, pixel_dimension)`` pairs
+       of length equal to the number of pixel dimensions in the extra coords.
+
+    """
+
+    @abc.abstractmethod
+    def __init__(self,
+                 *,
+                 wcs: BaseLowLevelWCS = None,
+                 mapping: Iterable[Tuple[int, int]] = None):
+        pass
+
+    @abc.abstractmethod
+    def add_coordinate(self,
+                       name: str,
+                       array_dimension: Union[int, Iterable[int]],
+                       lookup_table: Any,
+                       **kwargs):
+        """
+        Add a coordinate to this ``ExtraCoords`` based on a lookup table.
+
+        Parameters
+        ----------
+        name
+            The name for this coordinate(s).
+        array_dimension
+            The pixel dimension(s), in the array, to which this lookup table corresponds.
+        lookup_table
+            The lookup table.
+        """
+
+    @abc.abstractmethod
+    def keys(self) -> Iterable[str]:
+        """
+        The world axis names for all the coordinates in the extra coords.
+        """
+
+    @abc.abstractmethod
+    @property
+    def mapping(self) -> Iterable[Tuple[int, int]]:
+        """
+        The mapping between the array dimensions and pixel dimensions.
+
+        This is an iterable of ``(array_dimension, pixel_dimension)`` pairs
+        of length equal to the number of pixel dimensions in the extra coords.
+        """
+
+    @abc.abstractmethod
+    @property
+    def wcs(self) -> BaseHighLevelWCS:
+        """
+        A WCS object representing the world coordinates described by this ``ExtraCoords``.
+
+        .. note::
+            This WCS object does not map to the pixel dimensions of the array
+            associated with the `.NDCube` object. It has the number of
+            pixel dimensions equal to the number of inputs to the transforms to
+            get the world coordinates (normally equal to the number of world
+            coordinates). Therefore using this WCS directly might lead to some
+            confusing results.
+
+        """
+
+    @abc.abstractmethod
+    def __getitem__(self, item: Union[str, int, slice, Iterable[Union[str, int, slice]]]) -> "ExtraCoordsABC":
```

Comment:
Can slicing by `str` and `int`/`slice` really be suported in the one operation?  Or do you have to do either/or?

Reply:
it's not actually true you can slice with more than one string.

```suggestion
    def __getitem__(self, item: Union[str, int, slice, Iterable[Union[int, slice]]]) -> "ExtraCoordsABC":
```